typescript/introduction/tsc-variable-declaration/tsc-block-scoping.md

```diff
@@ -0,0 +1,98 @@
+---
+author: kapnobatai136
+
+type: normal
+
+category: must-know
+
+aspects:
+  - introduction
+
+---
+
+# Block-Scoping
+
+---
+## Content
+
+Let's talk a bit more about how the block scope works.
+
+We know that when a variable is defined using `let` or `const`, it's only accessible in the containing pair of curly braces, i.e. `{}`.
+
+Let's take another look at an example to understand this better:
+
+```ts
+function theBestApp() {
+  let getApp;
+
+  if (true) {
+    let app = 'Enki';
+    getApp = function() {
+      return app;
+    }
+  }
+
+  return getApp()
+}
+console.log(theBestApp());
+// Enki
+```
+
+Because we first define `getApp` and then we assign it a function, we won't get any errors.
+
+It should be noted that every iteration of a loop creates a new block-scope.
+
+```ts
+// the variable i is redeclared
```

Comment:
This is not true. In each iteration, the value of `i` is updated and since the value is primitive and because of closure the value of `i` in that iteration is printed. With a reference type you can see that its only reassignment in 3rd part of loop being triggered, and no redeclaration

```js
for (let i = 0, obj={a: 0}; i < 3; i++, obj.a=i) {
  setTimeout(function() { console.log(obj, i) }, 1000 * i)
};

// {a: 3} 0
// {a: 3} 1
// {a: 3} 2
```